scipy/cluster/hierarchy.py

```diff
@@ -1513,31 +1532,39 @@ def optimal_leaf_ordering(Z, y, metric='euclidean'):
     """
     xp = array_namespace(Z, y)
     Z = _asarray(Z, order='C', xp=xp)
-    is_valid_linkage(Z, throw=True, name='Z')
-
+    Z = xpx.lazy_wait_on(Z, is_valid_linkage(Z, throw=True, name='Z'))
     y = _asarray(y, order='C', dtype=xp.float64, xp=xp)
 
     if y.ndim == 1:
         distance.is_valid_y(y, throw=True, name='y')
     elif y.ndim == 2:
-        if (y.shape[0] == y.shape[1] and np.allclose(np.diag(y), 0) and
-                np.all(y >= 0) and np.allclose(y, y.T)):
-            warnings.warn('The symmetric non-negative hollow observation '
-                          'matrix looks suspiciously like an uncondensed '
-                          'distance matrix',
-                          ClusterWarning, stacklevel=2)
+        if (y.shape[0] == y.shape[1]):
+            y = xpx.lazy_warn(
+                y,
+                xpx.allclose(xp.diagonal(y), 0) & xp.all(y >= 0) & xpx.allclose(y, y.T),
+                'The symmetric non-negative hollow observation matrix looks '
+                'suspiciously like an uncondensed distance matrix',
+                ClusterWarning,
+                stacklevel=2,
+            )
         y = distance.pdist(y, metric)
-        y = xp.asarray(y)
     else:
         raise ValueError("`y` must be 1 or 2 dimensional.")
 
-    if not xp.all(xp.isfinite(y)):
-        raise ValueError("The condensed distance matrix must contain only "
-                         "finite values.")
+    y = xpx.lazy_raise(
+        y,
+        ~xp.all(xp.isfinite(y)),
+        ValueError("The condensed distance matrix must contain only finite values."),
+    )
 
-    Z = np.asarray(Z)
-    y = np.asarray(y)
-    return xp.asarray(_optimal_leaf_ordering.optimal_leaf_ordering(Z, y))
+    return xpx.lazy_apply(
+        _optimal_leaf_ordering.optimal_leaf_ordering,
+        Z, 
+        y,
+        shape=Z.shape,
```

Comment:
very minor point, but I wonder whether this pattern is more readable in general (when there is more than one argument to the callable), to separate the callable's params from those of `xpx.lazy_apply`
```suggestion
    return xpx.lazy_apply(
        _optimal_leaf_ordering.optimal_leaf_ordering,
        *(Z, y),
        shape=Z.shape,
```